software/python/state.py

```diff
@@ -0,0 +1,153 @@
+        self.cap = cap
+        self.robot = robot
+        self.opp_thresholder = opp_thresholder
+        self.pole_thresholder = pole_thresholder
+        self.ball_thresholder = ball_thresholder
+        self.ball_detector = ball_detector
+        self.pole_detector = pole_detector
+        self.set_state(State.BALL)
+        self.state_start_time = time.time()
+        self.map_to_max = map_to_max
+        self.orbit_left = orbit_left
+        self.orbit_right = orbit_right
+        self.delay_camera = delay_camera
+        self.go_forward = go_forward
+    
+
+    def img_processing(self):
+        frame, depth_frame = self.cap.get_frames()
+        self.frame_hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
+        self.depth_frame = depth_frame
+    
+
+    def state_handler(self):
+        self.img_processing()
+        if (self.get_state() == State.BALL):
+            self.state_ball()
+        elif (self.get_state() == State.OPP_BASKET):
+            self.state_go_to_opponent_basket()
+        elif (self.get_state() == State.THROW_GOAL):
+            self.state_throw_to_goal()
+        else:
+            raise Exception("Unexpected state")
+        
+        cv2.imshow("Aken", self.frame_hsv)
+
+
+    def get_state(self):
+        return self.state
+
+    
+    def set_state(self, state):
+        self.state = state
+        self.state_start_time = time.time()
+
+
+    def state_go_to_opponent_basket(self):
+        if time.time() - self.state_start_time > STATE_OVERFLOW_TIME:
+            self.set_state(State.BALL)
+        thresholded = self.create_thres_img_for_blob_detection(self.opp_thresholder)
+        cv2.imshow("Thresholded", thresholded)
+
+        kps = list(self.pole_detector.detect(thresholded))
+
+        self.frame_hsv = cv2.drawKeypoints(self.frame_hsv, kps, np.array([]), (0, 255, 255), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+        if len(kps) == 0:
+            self.robot.set_speed(8 / 32767)
+            self.robot.spin_left()
+        else:
+            self.robot.forward()
+    
+
+    def state_ball(self):
+        if time.time() - self.state_start_time > STATE_OVERFLOW_TIME:
+            self.set_state(State.OPP_BASKET)
+            self.state_start_time = time.time()
+        thresholded = self.create_thres_img_for_blob_detection(self.ball_thresholder)
+        cv2.imshow("Thresholded", thresholded)
+
+        keypoints = list(self.ball_detector.detect(thresholded))
+        keypoints.sort(key=lambda x: -x.size)  #  Biggest ball as first
+
+        self.frame_hsv = cv2.drawKeypoints(self.frame_hsv, keypoints, np.array([]), (0, 255, 0), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+        if len(keypoints) == 0:
+            self.robot.set_speed(8 / 32767)
+            self.robot.spin_left()
+        else:
+            kp = keypoints[0]
+            x_loc = kp.pt[0]
+            if kp.size < 45:                        
```

Comment:
Technically it is quite bad that you have image processing logic like pixel size filtering in the state logic. It should be handled in a base image processor with its set responsibilities